File: Heap/Python/K Closest Points to Origin.py
# Problem link: https://leetcode.com/problems/k-closest-points-to-origin/description/
# Time: O(n log n)
# Space: O(n)
class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        minHeap = []
        result = []

        for p1, p2 in points:
            heapq.heappush(minHeap, (math.sqrt(p1 ** 2 + p2 ** 2), p1, p2))
        
        while k > 0:
            detail = heapq.heappop(minHeap)
            result.append([detail[1], detail[2]])
            k -= 1
        
        return result

File: Heap/Java/K Closest Points to Origin.java
// Problem link: https://leetcode.com/problems/k-closest-points-to-origin/description/
// Time: O(n log n)
// Space: O(n)
class Solution {
    public int[][] kClosest(int[][] points, int k) {
        PriorityQueue<int[]> minHeap = new PriorityQueue<>((a, b) -> 
            Integer.compare(a[0] * a[0] + a[1] * a[1],
                            b[0] * b[0] + b[1] * b[1]));

        int[][] result = new int[k][2];

        for(int[] pair : points) {
            minHeap.offer(pair);
        }

        for(int idx = 0; idx < k; idx++) {
            int[] pair = minHeap.poll();
            result[idx][0] = pair[0];
            result[idx][1] = pair[1];
        }

        return result;
    }
}
